### ISSUE TITLE
大模型底层机制

### ISSUE THREAD

**Issue by laoyutong** [laoyutong/blog#65]
## token
模型不认识"文字"，它只认识数字。所以输入的任何内容——中文、英文、代码、标点符号——都会先被切成一小块一小块的"token"，每个 token 对应一个数字编号

大部分大模型用的分词算法叫 BPE（Byte Pair Encoding）：模型在训练之前，先统计大量文本中哪些字符组合出现频率最高，把高频组合合并成一个 token。英文训练数据比中文多得多，所以英文的合并更充分，一个词一个 token

中文经常要拆开来编码，导致了一个实际问题：同样的语义内容，中文比英文消耗更多的 token

## 生成方式
大模型不是"想好了再说"的，它是"边说边想"的：它先生成第 1 个 token，然后把第 1 个 token 加入输入，再生成第 2 个 token，然后把前 2 个 token 加入输入，再生成第 3 个……如此循环，直到生成一个"结束"标记
这个过程叫自回归生成（Autoregressive Generation）

对Agent开发的直接影响：
- 天然的流式响应
因为模型本来就是一个 token 一个 token 生成的，所以不需要等它全部说完再展示给用户。生成一个就发一个，这就是 SSE（Server-Sent Events）流式响应
- 工具调用的 JSON 要"攒完"才能解析
模型在调用工具的时候，会流式输出一堆 JSON 碎片：
```
{"na        ← 第一批 token
me": "rea   ← 第二批
d_file",    ← 第三批
"path":     ← 第四批
"/src/in    ← 第五批
dex.ts"}    ← 最后一批，JSON 才完整
```
- response prefill 能"操控"模型的行为
想让模型只能调用浏览器相关的工具，可以在 assistant 消息里预先填入，模型看到这个前缀，就只能在 browser_ 开头的工具名里选了
```
{"name": "browser_
```

## 注意力机制
Transformer 架构里最核心的内容，涉及Query、Key、Value三个角色
- Query查询：当前正在生成的 token 会问一个问题——我需要什么信息来决定下一个词？
- Key 索引：前面每一个 token 都有一个标签——我包含什么类型的信息？
- Value 内容：前面每一个 token 的实际内容——这是我的具体信息
生成新 token 的时候，模型拿当前 token 的 Query，去和前面所有 token 的 Key 做匹配。匹配度高的，就把对应的 Value 拿过来，加权混合，作为生成下一个 token 的依据

🌰 假设上下文是"小明昨天去了北京，今天他去了"，现在要生成下一个 token：
- 当前 token 的 Query 大概在问："谁去了哪里？"
- "小明"这个 token 的 Key 标签是"人名"，匹配度高 → 它的 Value 被拉过来
- "北京"的 Key 是"地名"，匹配度也挺高 → Value 也被拉过来
- "昨天"的 Key 是"时间词"，跟当前问题关系不大 → Value 的权重就低
最终模型综合了高权重的 Value 信息，可能生成"上海"（另一个城市）

所以问题是：每个新 token 都要拿自己的 Query 去和前面所有 token 的 Key 做匹配。n 个 token 就要做 n 次匹配。随着上下文越来越长，每个 token 能分到的"注意力预算"就越少。上下文里塞的东西越多，模型对每一条信息的"记忆力"就越差

## KV Cache
如果模型已经生成了 100 个 token，现在要生成第 101 个，它需要用到前 100 个 token 的 Key 和 Value，如果重新计算太浪费资源
所以模型会把之前算过的每个 token 的 Key 和 Value 缓存起来，这个缓存就叫 KV Cache（Key-Value Cache）

KV Cache 有一个关键限制：它是基于"前缀匹配"的，缓存只有在前缀完全一样的时候才能复用
- 合理🌰
模型只需要计算新增的 [用户消息2] 的部分
```
// 第一次对话
[System Prompt] + [用户消息1] + [模型回复1]
// 第二次对话
[System Prompt] + [用户消息1] + [模型回复1] + [用户消息2]
```
- 不合理🌰
第一个 token 就不一样了，后面所有的 KV Cache 全部作废，需要从头重新计算
```
当前时间：2026-04-02 08:00:00    ← 这个每次都变！
[其余的 System Prompt]
[用户消息1]
[模型回复1]
[用户消息2]
```

## 模型选词
1. 给所有候选词打分（Logits）
每生成一个新 token，模型会给词汇表里的每一个 token 打一个原始分数，表示"根据前面的上下文，这个 token 接下来出现的可能性有多大"
🌰：在"今天天气真"这个上下文后面，"好"这个 token 可能得分 5.2，"不"得分 3.1，"的"得分 1.8，而"跑步"可能只有 -2.3
2. 把分数变成概率（Softmax）
原始分数不好直接用，有正有负，也不知道总和是多少，所以需要把这些分数转换成概率分布，所有候选词的概率加起来等于 1。分数越高的词，概率越大，而且这个差距会被放大，高分词的概率会比它的原始分数暗示的还要高
🌰：经过 softmax 之后，"好"可能变成 65% 的概率，"不"15%，"的"8%，"跑步"0.01%
3. 根据概率采样（Sampling）
拿到概率分布后，模型从里面"抽签"选一个 token
temperature参数调的是 softmax 输出的概率分布的"尖锐程度"：
- temperature 接近 0：概率分布变得非常尖锐，高的特别高，低的也很低，最高分的那个 token 几乎 100% 被选中。模型输出变得确定、保守
- temperature = 1：正常的概率分布，有一定的随机性
- temperature > 1：概率分布变得更平坦（或者更平均），低分 token 也有机会被选中。模型输出变得更有创意，但也更容易胡说八道

### Logit Mask（约束解码）
既然模型是从概率分布里"抽签"的，那在抽签之前，可以直接把某些 token 的概率设为 0
🌰：希望模型输出 "yes" 或 "no"，那就把除了 "yes" 和 "no" 之外的所有 token 的 logit 设为负无穷（经过 softmax 后概率就是 0 了）。模型只能在 "yes" 和 "no" 之间选

### Structured Output（结构化输出）
在每一步生成的时候，用 logit mask 只允许当前上下文下语法合法的 token
🌰：刚输出了 {"name": 之后，下一个 token 只能是引号 " 开头的字符串
